compiler/noirc_evaluator/src/ssa/validation/mod.rs

```diff
@@ -0,0 +1,280 @@
+//! Validator that checks whether a function is well formed.
+//!
+//! It validates:
+//! - That the function contains exactly one return block.
+/// - That every checked signed addition or subtraction instruction is
+///   followed by a corresponding truncate instruction with the expected bit sizes.
+use fxhash::FxHashSet as HashSet;
+
+use crate::ssa::ir::instruction::TerminatorInstruction;
+
+use super::ir::{
+    function::Function,
+    instruction::{BinaryOp, Instruction, InstructionId},
+    types::{NumericType, Type},
+    value::ValueId,
+};
+
+/// Aside the function being validated, the validator maintains internal state
+/// during instruction visitation to track patterns that span multiple instructions.
+struct Validator<'f> {
+    function: &'f Function,
+    // State for truncate-after-signed-sub validation
+    // Stores: Option<(bit_size, result)>
+    signed_binary_op: Option<(u32, ValueId)>,
+}
+
+impl<'f> Validator<'f> {
+    fn new(function: &'f Function) -> Self {
+        Self { function, signed_binary_op: None }
+    }
+
+    fn start_function(&mut self) {
+        self.validate_single_return_block();
+        // Reset any state needed at start, if any
+        self.signed_binary_op = None;
```

Comment:
This looks a little strange. Is calling `start_function` multiple times an actual use case? I thought we call `new` and then `start_function`, so all state should be as it is by default. 

Maybe the iteration over the instructions could be moved into the `Validator`, to make it self contained. We clearly assume that we will visit all instructions and then call finish, not the other way around, and start in between them.

Reply:
Agree that this is a little odd. I'm fine with us doing multiple passes over each function in order to perform all of the various validations as this validation shouldn't be performance critical. Better to optimize for simplicity.

Reply:
Yeah agreed this is weird. I originally tried to make the validators re-use the same pass over the function and made this "start", "visit", "finish" logic. However, with the combined validator this isn't necessary and is leftover logic that has leaked here. I'm going to get rid of this and move the visiting of instructions to the `Validator`.